fix(custom-element): fix parent, observer and exposed issues when remove element and append it back after fully unmounted

lejunyang opened the pull request:
close https://github.com/vuejs/core/issues/12412

changes:
- extract observer code
- observe again and re-mount if it's resolved and no instance, new parent will be set on new mount
- exposed properties should be configurable, and we should delete them if element is fully unmounted, so that we can expose again in new mount

edison1105 commented:
Thank you for your contribution. I have some suggestions and questions.

> Click "remove el2" and "append el2"

I think the reason for the error here is that `_parent` and `_resolved` were not reset when disconnected. Therefore, the correct fix should be:
```js
// in disconnectedCallback
this._parent = undefined
this._resolved = false
```


> and then click "change el2 attr", its attribute is changed, but prop is not updated

The fix to this part are generally correct. However, I don't understand why [this line](https://github.com/vuejs/core/pull/12413#discussion_r1846041742) needs to be added.


> extract observer code

I think this change is no longer necessary if reset `_parent` and `_resolved` in `disconnectedCallback`

lejunyang commented on packages/runtime-dom/src/apiCustomElement.ts:
```diff
@@ -430,11 +444,14 @@ export class VueElement
-      } else if (__DEV__) {
-        warn(`Exposed property "${key}" already exists on custom element.`)
+      } else {
+        delete exposed[key] // delete it from exposed in case of deleting wrong exposed key when disconnected
```
This is to remove invalid exposed key. Let's say we have a element with prop `foo`, and we accidentally expose an object `foo`, it will be ignored as `foo` is already on this element. In disconnectedCallback, we'll delete exposed properties according to `exposed`, but deleting `foo` can cause an error as it's non-configurable prop.
Therefore, we should remove that key ahead

lejunyang commented:
@edison1105 Thanks for your suggestions! It's a good fix for current issues. 
However I found another issue regarding moving element. If we move a child element directly into other element(not waiting two nextTick), its parent context is not updated

I will check more on this and update issue and commit this tonight

lejunyang commented:
Actually setting `_resolved` to false can cause some duplicate executions, like `_applyStyles`

edison1105 commented:
> Actually setting `_resolved` to false can cause some duplicate executions, like `_applyStyles`

re-mount should also re-apply the style, right ?

lejunyang commented:
The reason `work with Teleport (shadowRoot: false)` test case fails is that: `my-y` is defined first, `my-p` is defined later, and it removes `my-p` and causes unmounting.
I think this is not code issue, it's test case issue. We should always define outer element first, especially when SSR, otherwise inner element can't find correct parent, so I updated this test case

lejunyang commented:
Updated, thanks for the info
Also I was correct, `this._resolved = false` will lead to `applyStyle` twice. `unmount` doesn't remove style nodes so we don't need to `applyStyle` again. I have added that in test case
So we just need to mount again instead of `resolveDef` again

wolandec commented:
Hey team any updates on this one?

HaNdTriX commented:
I just tested this Branch:

```sh
 pnpm add https://pkg.pr.new/vue@12413
 ```
 
 The fix solves this issue. Thanks @lejunyang
